manta-trusted-setup/src/ceremony/requests.rs

```diff
@@ -0,0 +1,260 @@
+
+//! Request Tools
+
+// use crate::{
+//     ceremony::{
+//         queue::{Identifier, Priority},
+//         signature::{self, SignatureScheme},
+//         CeremonyError,
+//     },
+//     mpc,
+// };
+// use core::{fmt::Debug, marker::PhantomData};
+// use serde::{Deserialize, Serialize};
+
+// /// Register Request
+// #[derive(Debug, Deserialize, Serialize)]
+// pub struct RegisterRequest<P>
+// where
+//     P: Identifier + Priority + signature::HasPublicKey,
+// {
+//     /// The Participant to register
+//     pub participant: P,
+// }
+
+// // TODO: Current SignatureScheme is subject to replay attacks.
+// impl<P, S> signature::Sign<S> for RegisterRequest<P>
+// where
+//     P: Identifier + Priority + signature::HasPublicKey<PublicKey = S::PublicKey>,
+//     S: SignatureScheme,
+//     P::PublicKey: signature::Sign<S>,
+// {
+//     #[inline]
+//     fn sign(
+//         &self,
+//         nounce: &S::Nounce,
+//         public_key: &S::PublicKey,
+//         private_key: &S::PrivateKey,
+//     ) -> Result<S::Signature, CeremonyError> {
+//         public_key.sign(nounce, public_key, private_key)
+//     }
+// }
+
+// impl<P, S> signature::Verify<S> for RegisterRequest<P>
+// where
+//     P: Identifier + Priority + signature::HasPublicKey<PublicKey = S::PublicKey>,
+//     S: SignatureScheme,
+//     P::PublicKey: signature::Verify<S>,
+// {
+//     #[inline]
+//     fn verify(
+//         &self,
+//         nounce: &S::Nounce,
+//         public_key: &S::PublicKey,
+//         signature: &S::Signature,
+//     ) -> Result<(), CeremonyError> {
+//         public_key.verify(nounce, public_key, signature)
+//     }
+// }
+
+// /// Request to join the contributor queue as `participant`.
+// #[derive(Debug, Deserialize, Serialize)]
+// pub struct JoinQueueRequest<P>
+// where
+//     P: Identifier + Priority,
+// {
+//     /// Participant
+//     participant: P,
+// }
+
+// /// Signed request to get the MPC state.
+// #[derive(Debug, Deserialize, Serialize)]
+// pub struct GetMpcRequest<P, V>
+// where
+//     P: Identifier,
+//     V: mpc::Verify,
+// {
+//     /// Participant
+//     pub participant: P,
+//     __: PhantomData<V>,
+// }
+
+// // TODO: need further discussion: with fixed `domain_tag`, if user leaks the signature during the trusted setup
+// // then someone can mock this `GetMpcRequest` and get the user's position.
+// impl<P, S, V> signature::Sign<S> for GetMpcRequest<P, V>
+// where
+//     P: Identifier + signature::HasPublicKey<PublicKey = S::PublicKey>,
+//     S: SignatureScheme,
+//     V: mpc::Verify,
+//     P::PublicKey: signature::Sign<S>,
+// {
+//     #[inline]
+//     fn sign(
+//         &self,
+//         nounce: &S::Nounce,
+//         public_key: &S::PublicKey,
+//         private_key: &S::PrivateKey,
+//     ) -> Result<S::Signature, CeremonyError> {
+//         // sign the public key
+//         public_key.sign(nounce, public_key, private_key)
```

Comment:
Should we sign both the public key and other information such as identifier and priority?